contracts/BC_fusion/TokenRecoverPortal.sol

```diff
@@ -0,0 +1,298 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./interface/ITokenHub.sol";
+import "./interface/ITokenRecoverPortal.sol";
+import "./System.sol";
+import "./lib/Utils.sol";
+
+/**
+ * @title TokenRecoverPortal is used to recover the token from BC users.
+ * @dev This is designed for the BC users to recover the token from TokenHub.
+ * The BC chain will stop and generate a merkle tree root after BC-fusion plan was started.
+ * The BC users can recover the token from TokenHub after the merkle tree root is generated.
+ * For more details, please refer to the BEP-299(https://github.com/bnb-chain/BEPs/pull/299).
+ */
+contract TokenRecoverPortal is System, ReentrancyGuardUpgradeable {
+    using Utils for string;
+    using Utils for bytes;
+
+    /*----------------- constants -----------------*/
+    string public constant SOURCE_CHAIN_ID = "Binance-Chain-Tigris";
+
+    /*----------------- storage -----------------*/
+    address public approvalAddress;
+    bytes32 public merkleRoot;
+    bool public merkleRootAlreadyInit;
+
+    // recoveredMap is used to record the recovered token.
+    mapping(bytes32 => bool) private recoveredMap;
+
+    // assetProtector is the address that is allowed to pause the #recover.
+    address public assetProtector;
+    // paused is used to pause the #recover.
+    bool private _paused;
+
+    modifier whenNotPaused() {
+        if (_paused) revert TokenRecoverPortalPaused();
+        _;
+    }
+
+    modifier onlyAssetProtector() {
+        if (msg.sender != assetProtector) revert OnlyAssetProtector();
+        _;
+    }
+
+    modifier merkelRootReady() {
+        if (!merkleRootAlreadyInit) revert MerkleRootNotInitialize();
+        if (merkleRoot == bytes32(0)) revert MerkleRootNotInitialize();
+        _;
+    }
+
+    modifier approvalAddressInit() {
+        if (approvalAddress == address(0)) revert ApprovalAddressNotInitialize();
+        _;
+    }
+
+    /*----------------- errors -----------------*/
+    // @notice signature: 0x3e493100
+    error AlreadyRecovered();
+    // @notice signature: 0x09bde339
+    error InvalidProof();
+    // @notice signature: 0xad60149e
+    error InvalidApprovalSignature();
+    // @notice signature: 0x8152ea1b
+    error InvalidOwnerPubKeyLength();
+    // @notice signature: 0xbc97af2e
+    error InvalidOwnerSignatureLength();
+    // @notice signature: 0xf36660de
+    error MerkleRootAlreadyInitiated();
+    // @notice signature: 0x57407297
+    error MerkleRootNotInitialize();
```

Comment:
`MerkleRootNotInitialize` -> `MerkleRootNotInitialized` to align with others

Reply:
Done